<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head><title>R: Metabolic Analysis using Queries</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>
<h1> Metabolic Analysis using Queries
<img class="toplogo" src="../../../doc/html/Rlogo.svg" alt="[R logo]" />
</h1>
<hr/>
<div style="text-align: center;">
<a href="../../../doc/html/packages.html"><img class="arrow" src="../../../doc/html/left.jpg" alt="[Up]" /></a>
<a href="../../../doc/html/index.html"><img class="arrow" src="../../../doc/html/up.jpg" alt="[Top]" /></a>
</div><h2>Documentation for package &lsquo;MetQy&rsquo; version 1.0.1</h2>

<ul><li><a href="../DESCRIPTION">DESCRIPTION file</a>.</li>
</ul>

<h2>Help Pages</h2>


<table width="100%">
<tr><td style="width: 25%;"><a href="analysis_genomes_module_output.html">analysis_genomes_module_output</a></td>
<td>Process the output generated by 'query_genomes_to_modules()'.</td></tr>
<tr><td style="width: 25%;"><a href="analysis_pca_mean_distance_calculation.html">analysis_pca_mean_distance_calculation</a></td>
<td>Given a set of coordinates, calculate the mean distance between all points.</td></tr>
<tr><td style="width: 25%;"><a href="analysis_pca_mean_distance_grouping.html">analysis_pca_mean_distance_grouping</a></td>
<td>Given a set of coordinates and a grouping variable, the mean distance is calculated for each group.</td></tr>
<tr><td style="width: 25%;"><a href="data_example_ECnumbers_vector.html">data_example_ECnumbers_vector</a></td>
<td>Example of a vector containing Enzyme Commission (EC) numbers.</td></tr>
<tr><td style="width: 25%;"><a href="data_example_KOnumbers_vector.html">data_example_KOnumbers_vector</a></td>
<td>Example of a vector containing KEGG Ortholog (KO) identifiers</td></tr>
<tr><td style="width: 25%;"><a href="data_example_multi_ECs_KOs.html">data_example_multi_ECs_KOs</a></td>
<td>Example of a data frame with multiple datasets.</td></tr>
<tr><td style="width: 25%;"><a href="data_module_shortName_mapping.html">data_module_shortName_mapping</a></td>
<td>Data frame used to populate the ahort name in module_reference_table</td></tr>
<tr><td style="width: 25%;"><a href="misc_axisRound.html">misc_axisRound</a></td>
<td>Find best value to round axis to.</td></tr>
<tr><td style="width: 25%;"><a href="misc_check_duplicate_names.html">misc_check_duplicate_names</a></td>
<td>Check for duplicated strings and append letter code to distinguish them.</td></tr>
<tr><td style="width: 25%;"><a href="misc_create_labels.html">misc_create_labels</a></td>
<td>Creats labels to add to a character vector to make all entries unique</td></tr>
<tr><td style="width: 25%;"><a href="misc_evaluate_block.html">misc_evaluate_block</a></td>
<td>Evaluate a KEGG module block definition.</td></tr>
<tr><td style="width: 25%;"><a href="misc_geneVector_module.html">misc_geneVector_module</a></td>
<td>Map a list of EC or K numbers to KEGG modules.</td></tr>
<tr><td style="width: 25%;"><a href="misc_module_definition_block_EC.html">misc_module_definition_block_EC</a></td>
<td>KEGG module definition processing: subblocks</td></tr>
<tr><td style="width: 25%;"><a href="misc_module_definition_check.html">misc_module_definition_check</a></td>
<td>KEGG module definition processing: block formatting.</td></tr>
<tr><td style="width: 25%;"><a href="misc_module_definition_optional.html">misc_module_definition_optional</a></td>
<td>KEGG module definition processing: KEGG ortholog exclusion</td></tr>
<tr><td style="width: 25%;"><a href="misc_module_subgroup_indexing.html">misc_module_subgroup_indexing</a></td>
<td>Subgroup check - formatting KEGG module definition.</td></tr>
<tr><td style="width: 25%;"><a href="parseKEGG_compound.html">parseKEGG_compound</a></td>
<td>Parse the KEGG compound database</td></tr>
<tr><td style="width: 25%;"><a href="parseKEGG_enzyme.html">parseKEGG_enzyme</a></td>
<td>Parse the KEGG enzyme database</td></tr>
<tr><td style="width: 25%;"><a href="parseKEGG_execute_all.html">parseKEGG_execute_all</a></td>
<td>Execute all parseKEGG parent functions to format KEGG databases into data frames</td></tr>
<tr><td style="width: 25%;"><a href="parseKEGG_file.html">parseKEGG_file</a></td>
<td>Parse any KEGG file without extension</td></tr>
<tr><td style="width: 25%;"><a href="parseKEGG_file.list.html">parseKEGG_file.list</a></td>
<td>Parse any '.list' KEGG file</td></tr>
<tr><td style="width: 25%;"><a href="parseKEGG_genome.html">parseKEGG_genome</a></td>
<td>Parse the KEEG genome database</td></tr>
<tr><td style="width: 25%;"><a href="parseKEGG_ko.html">parseKEGG_ko</a></td>
<td>Parse the KEGG orthology (KO) database</td></tr>
<tr><td style="width: 25%;"><a href="parseKEGG_ko_enzyme.html">parseKEGG_ko_enzyme</a></td>
<td>Map KEGG orthology (KO) to Enzyme Commission (EC) numbers</td></tr>
<tr><td style="width: 25%;"><a href="parseKEGG_ko_reaction.html">parseKEGG_ko_reaction</a></td>
<td>Map KEGG orthologs (KOs) to Reaction IDs</td></tr>
<tr><td style="width: 25%;"><a href="parseKEGG_module.html">parseKEGG_module</a></td>
<td>Parse the KEGG module database</td></tr>
<tr><td style="width: 25%;"><a href="parseKEGG_process_KEGG_taxonomy.html">parseKEGG_process_KEGG_taxonomy</a></td>
<td>Retrieve the taxonomy listed as part of the KEGG genome database.</td></tr>
<tr><td style="width: 25%;"><a href="parseKEGG_reaction.html">parseKEGG_reaction</a></td>
<td>Parse KEGG reaction database</td></tr>
<tr><td style="width: 25%;"><a href="plot_heatmap.html">plot_heatmap</a></td>
<td>Plot a heatmap</td></tr>
<tr><td style="width: 25%;"><a href="plot_scatter.html">plot_scatter</a></td>
<td>Scatter plot</td></tr>
<tr><td style="width: 25%;"><a href="plot_scatter_byFactors.html">plot_scatter_byFactors</a></td>
<td>Scatter plot with overlapping factors/groups.</td></tr>
<tr><td style="width: 25%;"><a href="plot_variance_boxplot.html">plot_variance_boxplot</a></td>
<td>Make a boxplot</td></tr>
<tr><td style="width: 25%;"><a href="query_genes_to_genome.html">query_genes_to_genome</a></td>
<td>Find the genome(s) that express a set of genes.</td></tr>
<tr><td style="width: 25%;"><a href="query_genes_to_modules.html">query_genes_to_modules</a></td>
<td>Given a set of genes, find the modules they are involved in.</td></tr>
<tr><td style="width: 25%;"><a href="query_genomes_to_modules.html">query_genomes_to_modules</a></td>
<td>Evaluates the KEGG modules presence given genome information.</td></tr>
<tr><td style="width: 25%;"><a href="query_missingGenes_from_module.html">query_missingGenes_from_module</a></td>
<td>Identify missing genes from a KEGG module</td></tr>
<tr><td style="width: 25%;"><a href="query_modules_to_genomes.html">query_modules_to_genomes</a></td>
<td>Find a KEGG genome that has a given KEGG module complete</td></tr>
</table>
</body></html>
