#' Parse the KEGG orthology (KO) database
#'
#' Read and format the KEGG orthology (KO) database (containing ortholog (gene) information) text file into a reference table.
#'
#' @details
#' The columns are automatically generated by the \code{parseKEGG_file} function into variables,
#' which are further formatted specifically for the KEGG ortholog database.
#'
#' The text file used is "\code{KEGG_path}/genes/ko/ko".
#'
#' It decompresses "\code{KEGG_path}/genes/ko.tar.gz" if needed.
#'
#' @param KEGG_path     - string pointing to the location of the KEGG database parent folder. The path to the required file is contained within the function.
#'
#' @param outDir        - optional. String pointing to the output folder. Default ("output/").
#'
#' @param verbose       - logical. Should progress be printed to the screen? Default (\code{TRUE}).
#'
#' @param ...       - further arguments for \code{parseKEGG_file()}.
#'
#' @return Generates ko_reference_table (.txt & .rda; saved to \code{'outDir'}) and returns
#' a data frame with as many rows as entries and the following columns (or variables):
#' \preformatted{
#'
#' 	    *In all instances, multiple entries in a given column are separated by '[;]'.
#' }
#'
#' @examples
#' KEGG_path <- "~/KEGG" # MODIFY TO KEGG PARENT FOLDER!
#' # The parent folder should contain the following (KEGG FTP structure):
#' #	brite/
#' #	genes/
#' #	ligand/
#' #	medicus/
#' #	module/
#' #	pathway/
#' #	README.kegg
#' #	RELEASE
#' #	xml/
#'
#' ko_reference_table <- parseKEGG_ko(KEGG_path)
#'  # A .txt file (tab separated) is written to output/ (relative to current working directory)
#'
#' @seealso \link{parseKEGG_file}
#'
#' @export


parseKEGG_ko <- function(KEGG_path, outDir = "output", verbose = T, ...){

  ####  MANAGE INPUT ----
  # CHECKS
  stopifnot(is.character(KEGG_path),length(KEGG_path)==1,dir.exists(KEGG_path))
  stopifnot((is.character(outDir)&&length(outDir)==1)||is.null(outDir))
  stopifnot(is.logical(verbose))

  # FORMAT PATHS
  KEGG_path <- gsub("/$","",KEGG_path) # end file path WITHOUT ending "/"
  if(!is.null(outDir)) outDir    <- gsub("/$","",outDir)

  # OUTPUT FOLDER
  if(!is.null(outDir)) if(!dir.exists(outDir)) dir.create(outDir)

  ####  PROCESS FILE ----
  if(verbose) cat("\tko processing...",fill = T)
  start <- Sys.time()

  ko_file  <- paste(KEGG_path,"/genes/ko/ko",sep="")

  ## UNTAR FILE
  if(!file.exists(paste(KEGG_path,"/genes/ko/ko",sep=""))){
    if(verbose) cat("\n\t\tDecompressing ko file...",fill = T)
    untar(paste(KEGG_path,"/genes/ko.tar.gz",sep=""),files = "ko/ko",exdir = paste(KEGG_path,"/genes/",sep=""))
  }

  ## PROCESS FILE INTO DATA FRAME
  # ko_reference_table <- parseKEGG_file('/Volumes/OSSLAB/mount_KEGG/genes/ko/ko')
  ko_reference_table <- parseKEGG_file(ko_file, verbose = verbose,...)

  ####  FORMAT TABLE                    ----
  colnames(ko_reference_table)[1] <- "ID"
  ko_reference_table$ID <- gsub(" KO","",ko_reference_table$ID)

  ####  WRITE TABLE  (.txt, .rda) ----
  ko_reference_table_file <- paste(outDir,"/ko_reference_table.txt",sep="")

  if(!is.null(outDir)){
    write.table(ko_reference_table,file = ko_reference_table_file,sep = "\t",row.names = F,quote = F)
    save(ko_reference_table,file = gsub(".txt",".rda",ko_reference_table_file))
  }
  if(verbose) cat("\t Completed in",format(difftime(Sys.time(),start,units = "mins"),digits = 4),"\n",fill = T)

  ### RETURN ---
  return(ko_reference_table)

}

# UNUSED!
# #### ADD ADDITIONAL INFORMATION: ORGANISM DEPENDENT ----
# # e.g.
# aaa_kos <- read.delim('/Volumes/OSSLAB/mount_KEGG/genes/organisms/aaa/T01445.kff',
#                       header = F,stringsAsFactors =F)
#
# names(aaa_kos) <- c(
#   'KEGG_GENES_ID', # identifier (in the form of org:gene)
#   'Feature_key',   # such as CDS, tRNA, rRNA, ncRNA, lncRNA, miRNA, and gene (usually meaning pseudogene)
#   'aa_seq_length', # Amino acid sequence length
#   'nt_seq_length', #Nucleotide sequence length
#   'Geonomic_position',
#   'NCBI_GeneID',
#   'NCBI_gi',
#   'Gene_name',
#   'Definition_original_DB',
#   'KO_identifier',# (K number)
#   'KO_definition'
# )
