#' Parse the KEEG genome database
#'
#' Read and format the KEGG genome database text file and the organism information and format it into a reference table.
#'
#' @details
#' The columns are automatically generated by the \code{parseKEGG_file} function into variables,
#' which are further formatted specifically for the KEGG genome database.
#'
#' The main text file used is "\code{KEGG_path}/genes/genome/genome".
#'
#' It decompresses "\code{KEGG_path}/genes/genome.tar.gz" if needed.
#'
#' If \code{addECs} and/or \code{addKOs} are set to \code{TRUE}, the KEGG genome 3-4 letter code identifier
#' are used to retrieve the enzyme and KEGG ortholog content for each KEGG genome, respectively.
#'
#' @param KEGG_path     - string pointing to the location of the KEGG database parent folder. The path to the required file(s) is contained within the function.
#'
#' @param outDir        - optional. String pointing to the output folder. Default ("output/").
#'
#' @param addKOs        - logical. Should the list of K numbers be retrieved for each organism? Default (\code{TRUE}).
#'
#' @param addECs        - logical. Should the list of ECs be retrieved for each organism? Default (\code{TRUE}).
#'
#' @param includeViruses - logical. Should viral genomes be included? Default (\code{FALSE}).
#'
#' @param verbose       - logical. Should progress be printed to the screen? Default (\code{TRUE}).
#'
#' @param ...       - further arguments for \code{parseKEGG_file()}.
#'
#' @return Generates genome_reference_table (.txt & .rda; saved to \code{'outDir'}) and returns
#' a data frame with as many rows as entries and the following columns (or variables):
#' \preformatted{
#' 	 (1) ID                   - KEGG genome   identifier (T0 number; e.g. "T00001");
#' 	 (2) ORG_ID               - KEGG organism identifier (3 or 4 letter code; e.g. "hin");
#' 	 (3) STATUS               - genome sequence status (e.g. "Complete Genome");
#' 	 (4) NAME                 - various identifiers (e.g. "hin, HAEIN, 71421");
#' 	 (5) ORGANISM             - organism name (Genus species sp);
#' 	 (6) ANNOTATION           - type of annotation (one of "manual", "KOALA" or  "none");
#' 	 (7) TAXONOMY;             (8)  DATA_SOURCE;
#' 	 (9) ORIGINAL_DB          - original database;
#' 	 (10) KEYWORDS;            (11) DISEASE;
#' 	 (12) COMMENT;             (13) CHROMOSOME;
#' 	 (14) STATS_N_NUCLEOTIDES - statistics, number of nucleotides;
#' 	 (15) STATS_N_GENES_PROT  - statistics, number of protein-encoding genes;
#' 	 (16) STATS_N_GENES_RNA   - statistics, number of RNA-encoding genes;
#' 	 (17) REFERENCE           - reference(s) for study from which genomic sequence was
#' 	                             derived;
#' 	 (18) PLASMID;             (19) DBLINKS;
#' 	 (20) KOs                 - concatenated string of KEGG Orthologs (K numbers;
#' 	                             e.g. "K00001");
#' 	 (21) ECs                 - concatenated string of Enzyme Classification (EC) numbers
#' 	                             (e.g. "1.1.1.1").
#'
#' 	    *In all instances, multiple entries in a given column are separated by '[;]'.
#' }
#'
#' @examples
#' KEGG_path <- "~/KEGG" # MODIFY TO KEGG PARENT FOLDER!
#' # The parent folder should contain the following (KEGG FTP structure):
#' #	brite/
#' #	genes/
#' #	ligand/
#' #	medicus/
#' #	module/
#' #	pathway/
#' #	README.kegg
#' #	RELEASE
#' #	xml/
#'
#' genome_reference_table <- parseKEGG_genome(KEGG_path)
#'  # A .txt file (tab separated) is written to output/ (relative to current working directory)
#'
#' @seealso \link{parseKEGG_file}
#'
#' @export

################################ STEPS ########################################################
# 1.    MANAGE INPUT  - format paths and check/make outDir
# 2.    PROCESS FILE  - path structure to file self-contained as per KEGG data structure
#         Uses - parseKEGG_file() & 'KEGG_path/genes/genome/genome'
# 3.    FORMAT TABLE
# 3.1     ENTRY (column 1) changed to ID and contains only taxonomy numbers (T\\d{5})
# 3.2     STATUS column - distinguish between "Complete Genome" OR "Viral Genome" (as of 27/2/2017; more classes possible with KEGG updates)
# 3.3     Rename DEFINITION to ORGANISM
# 3.4     Option to remove Viral Genomes
#           As per argument: 'includeViruses'
# 4.    ADD KO & EC INFO
#           As per arguments: 'addECs' and 'addKOs'
# 4.1     Loop through genes/genome/organisms/ directories
# 5.    WRITE TABLE  (.txt, .rda)
##############################################################################################

parseKEGG_genome <- function(KEGG_path, outDir = "output",addKOs = T,addECs = T, includeViruses = F, verbose = T,...){

  ####  MANAGE INPUT ----
  # CHECKS
  stopifnot(is.character(KEGG_path),length(KEGG_path)==1,dir.exists(KEGG_path))
  stopifnot((is.character(outDir)&&length(outDir)==1)||is.null(outDir))
  stopifnot(is.logical(addECs),is.logical(addKOs),is.logical(includeViruses),is.logical(verbose))

  # FORMAT PATHS
  KEGG_path <- gsub("/$","",KEGG_path) # end file path WITHOUT ending "/"
  if(!is.null(outDir)) outDir    <- gsub("/$","",outDir)

  # OUTPUT FOLDER
  if(!is.null(outDir)) if(!dir.exists(outDir)) dir.create(outDir)

  ####  PROCESS FILE ----
  if(verbose) cat("\tgenome processing...",fill = T)
  start <- Sys.time()

  org_file  <- paste(KEGG_path,"/genes/genome/genome",sep="")

  ## UNTAR FILE
  if(!file.exists(paste(KEGG_path,"/genes/genome/genome",sep=""))){
    if(verbose) cat("\n\t\tDecompressing genome file...",fill = T)
    untar(paste(KEGG_path,"/genes/genome.tar.gz",sep=""),files = "genome/genome",exdir = paste(KEGG_path,"/genes/",sep=""))
  }

  ## PROCESS FILE INTO DATA FRAME
  genome_reference_table <- parseKEGG_file(org_file,verbose = verbose,...)

  ####  FORMAT TABLE                    ----
  # Include 'status' - either "Complete Genome" OR "Viral Genome"
  STATUS <- gsub("T\\d{5} ","",genome_reference_table$ENTRY)

  # USE ONLY T\\d{5} and RENAME 'ENTRY' TO 'ID'
  genome_reference_table$ENTRY      <- gsubfn::strapplyc(pattern = "T\\d{5}",genome_reference_table$ENTRY,simplify = T)
  # names(genome_reference_table)[1]  <- "ID"
  # genome_reference_table$ID         <- as.character(genome_reference_table$ID)

  # RENAME 'DEFINITION' TO 'ORGANISM'
  names(genome_reference_table)[which(names(genome_reference_table)=="DEFINITION")] <- "ORGANISM"

  # KEEP VIRAL GENOMES?
  if(!includeViruses) if(length(grep("Viral",STATUS))>0){
    genome_reference_table  <- genome_reference_table[-grep("Viral",STATUS),]
    STATUS                  <- STATUS[-grep("Viral",STATUS)]
  }

  # RETRIEVE LETTER CODE FOR THE GENOMES (used to save info in folder structure,e.g. eco - E. coli - )
  ORG_ID <- gsubfn::strapplyc(genome_reference_table$NAME,pattern = "([a-z]+)[,]",simplify = T)
  if(is.list(ORG_ID)){
    for(L in 1:length(ORG_ID)) if(!length(ORG_ID[[L]])) ORG_ID[[L]] <- ""
    ORG_ID <- as.character(t(as.data.frame(ORG_ID,stringsAsFactors = F)))
  }

  ## INCLUDE 'ORG_ID' and 'STATUS' IN DATA FRAME
  genome_reference_table <- cbind(genome_reference_table[,1],ORG_ID,STATUS,genome_reference_table[,2:ncol(genome_reference_table)])

  # USE ONLY T\\d{5} and RENAME 'ENTRY' TO 'ID'
  names(genome_reference_table)[1]  <- "ID"

  # FORMAT DATA TYPE
  genome_reference_table$ID         <- as.character(genome_reference_table$ID)
  genome_reference_table$ORG_ID     <- as.character(genome_reference_table$ORG_ID)
  genome_reference_table$STATUS     <- as.character(genome_reference_table$STATUS)

  # STATISTICS ----
  STATISTICS <- data.frame("STATS_N_NUCLEOTIDES" = numeric(nrow(genome_reference_table)),
                           "STATS_N_GENES_PROT"  = numeric(nrow(genome_reference_table)),
                           "STATS_N_GENES_RNA"   = numeric(nrow(genome_reference_table)),
                           stringsAsFactors = F)
  for(S in 1:nrow(genome_reference_table)){
    patterns <- c("Number of nucleotides: (\\d+)","Number of protein genes: (\\d+)","Number of RNA genes: (\\d+)")

    for(P in 1:length(patterns)){
      extracted <- gsubfn::strapplyc(genome_reference_table$STATISTICS[S],pattern = patterns[P],simplify = T)
      STATISTICS[S,P] <- ifelse(!is.list(extracted),extracted,"NA")
    }
  }

  # ADD TO TABLE
  stats_index             <- which(names(genome_reference_table)=="STATISTICS")
  genome_reference_table  <- cbind(genome_reference_table[,1:(stats_index-1)],STATISTICS,genome_reference_table[,(stats_index+1):ncol(genome_reference_table)])

  ####  ADD KOs & ECs INFO ----
  if(addECs||addKOs){
    if(verbose) cat("\n\t\tRetrieving individual organism's EC and KO numbers", fill = T)
    genome_reference_table$KOs <- ""
    genome_reference_table$ECs <- ""


    #### LOOP THROUGH ORGANISMS ----
    for(i in 1:nrow(genome_reference_table)){

      if(verbose) if(i==nrow(genome_reference_table)||i%%100==0||i==1)   cat("\t",i,"\tof",nrow(genome_reference_table),"genomes", fill = T)

      ORG <- genome_reference_table$ORG_ID[i]
      # SKIP ORGANISMS WITHOUT LETTER CODE (e.g. eco)
      if(ORG=="") next

      # ENZYMES
      ORG_dir <- paste(KEGG_path,"/genes/organisms/",ORG,"/",sep="")
      if(dir.exists(ORG_dir)){
        # EXTRACT EC INFO
        if(addECs){
          # EXTRACT FILE

          untar(paste(KEGG_path,"/genes/organisms/",ORG,"/",ORG,"_link.tar.gz",sep=""),
              files = paste(ORG,"_enzyme.list",sep=""),
              exdir = paste(KEGG_path,"/genes/organisms/",ORG,"/",ORG,"_link/",sep=""))

          enzyme_file <- paste(KEGG_path,"/genes/organisms/",ORG,"/",ORG,"_link/",ORG,"_enzyme.list",sep="")
          if(file.exists(enzyme_file)){
            enzymes     <- read.delim(enzyme_file,header = F)
            genome_reference_table$ECs[i] <- paste(gsub("ec:","",sort(unique(enzymes[,2]))),collapse = ";")
          }
        }

        # EXTRACT KEGG ortholog (KO) INFO
        if(addKOs){
          # EXTRACT FILE
          untar(paste(KEGG_path,"/genes/organisms/",ORG,"/",ORG,"_link.tar.gz",sep=""),
                files = paste(ORG,"_ko.list",sep=""),
                exdir = paste(KEGG_path,"/genes/organisms/",ORG,"/",ORG,"_link/",sep=""))

          ko_file <- paste(KEGG_path,"/genes/organisms/",ORG,"/",ORG,"_link/",ORG,"_ko.list",sep="")
          if(file.exists(ko_file)){
            kos     <- read.delim(ko_file,header = F)
            genome_reference_table$KOs[i] <- paste(gsub("ko:","",sort(unique(kos[,2]))),collapse = ";")
          }
        }

        # DELETE EXTRACTED FOLDER
        unlink(paste(KEGG_path,"/genes/organisms/",ORG,"/",ORG,"_link/",sep=""),recursive = TRUE)
      } # end 'if dir.exists'
    } # end ORG loop
  } # end loop to add ECs & KOs

  ####  WRITE TABLE  (.txt, .rda) ----
  genome_reference_table_file <- paste(outDir,"/genome_reference_table.txt",sep="")

  if(!is.null(outDir)){
    write.table(genome_reference_table,file = genome_reference_table_file,sep = "\t",row.names = F,quote = F)
    save(genome_reference_table,file = gsub(".txt",".rda",genome_reference_table_file))
  }

  if(verbose) cat("\t Completed in",format(difftime(Sys.time(),start,units = "mins"),digits = 4),"\n",fill = T)

  ### RETURN ---
  return(genome_reference_table)
}
